runtime/src/bank.rs

```diff
@@ -626,12 +627,73 @@ impl Bank {
             &sysvar::rewards::create_account(1, validator_point_value, storage_point_value),
         );
 
+        let validator_rewards = self.pay_validator_rewards(validator_point_value);
+
         self.capitalization.fetch_add(
-            (validator_rewards + storage_rewards) as u64,
+            validator_rewards + storage_rewards as u64,
             Ordering::Relaxed,
         );
     }
 
+    /// iterate over all stakes, redeem vote credits for each stake we can
+    ///   successfully load and parse, return total payout
+    fn pay_validator_rewards(&self, validator_point_value: f64) -> u64 {
+        let stake_history = self.stakes.read().unwrap().history().clone();
```

Comment:
Why the clone?

Reply:
the self.store() below needs a write lock on self.stakes, so if I hold a ref to the stake_history, I hold a read lock and get a dead lock

Reply:
Can't inline it?  Maybe I'm missing a loop that makes that impractical.

```rust
 stake.calculate_rewards(
                                            validator_point_value,
                                            &vote_state,
                                            Some(&self.stakes.read().unwrap().history()),
                                        )
```

Reply:
that's exactly what I had that caused deadlock ;)